project/src/main/java/com/google/autograder/data/Answer.java

```diff
@@ -4,40 +4,40 @@
 import com.google.appengine.api.datastore.KeyFactory;
 
 /** Class containing answer data. */
-public class Answer {
-
-  public String filePath;
-  public String parsedAnswer;
-  public int score;
-  public String assignmentKey;
-  public String questionKey;
-  public String key;
-  public String groupKey;
-
-  public Answer(String filePath, String parsedAnswer, int score, String assignmentKey, String questionKey, Key key) {
-    this.filePath = filePath;
-    this.parsedAnswer = parsedAnswer;
-    this.score = score;
-    this.assignmentKey = assignmentKey;
-    this.questionKey = questionKey;
-    this.key = KeyFactory.keyToString(key);
-    this.groupKey = "";
-  }
-
-  public String getKeyString() {
-      return this.key;
-  }
-
-  public Key getKey() {
-      return KeyFactory.stringToKey(this.key);
-  }
-
-  public void addGroup(Key newGroupKey) {
-      this.groupKey = KeyFactory.keyToString(newGroupKey);
-  }
-
-  public String getGroupKey() {
-      return this.groupKey;
-  }
+public final class Answer {
+
+    public int score;
```

Comment:
These should be final as well. It looks like `addGroup` is always called right after constructing an Answer, so `groupKey` can also be final and passed in the constructor